### PR TITLE
No longer rely on static state in ConfigurationReader

### PR DESCRIPTION
**Pull request by tsimbalar**
We had a commented out test that failed in a non deterministic way, because of `static` state. 

This PR removes that static state altogether, and *un-skips* that test.

As part of this PR I also did quite a bit of refactoring as I was trying to wrap my head about its behavior :). I also tried to move some code out of `ConfigurationReader` which is starting to grow big.

- make only one constructor of `ConfigurationReader` public to make it clearer that we always expect a `ConfigSection` and optionally a full `IConfiguration`.
- introduce the `AssemblyFinder` abstract class that takes care of locating assemblies from `DependencyContext` or from DLL-scanning, and make it clearer that we try to automatically guess which one to use when not explicitly specified.
- move all the code related to locating assemblies under the `*.Assemblies` namespace
- introduce `ResolutionContext` that keeps track of the values that can be useful when calling configuration methods (Level switches, the actual `IConfiguration` object) that is passed around when processing sub-sections of the config file.
- reduce the number of parameters of the `ConfigurationReader` class
- centralize the case-insensitive match strategy used when matching settings values to method parameters
- align naming with the usual serilog projects conventions

I did get a bit carried away, and the diffs ends up being quite big and possibly hard to review. If you think it is necessary, I can try and split this PR into multiple commits.

fixes https://github.com/serilog/serilog-settings-configuration/issues/147

### CODE REVIEW

**Comment by nblumhardt** [comment]
Looks awesome; not sure how the static state there slipped past me, I am not in this headspace enough right now to trust myself to review this PR, but LGTM on the surface! :-) .. Hoping someone with a little more time spare can go through it properly...

File: src/Serilog.Settings.Configuration/Settings/Configuration/Assemblies/DllScanningAssemblyFinder.cs [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+using System.Linq;
+using System.Reflection;
+
+namespace Serilog.Settings.Configuration.Assemblies
+{
+    sealed class DllScanningAssemblyFinder : AssemblyFinder
+    {
+        public override IReadOnlyList<AssemblyName> FindAssembliesContainingName(string nameToFind)
+        {
+            var query = from outputAssemblyPath in System.IO.Directory.GetFiles(AppDomain.CurrentDomain.BaseDirectory, "*.dll")
```
[the remark this answers is not in the document]
Comment (skomis-mm) [reply src/Serilog.Settings.Configuration/Settings/Configuration/Assemblies/DllScanningAssemblyFinder.cs]: I'm ok 👍

[the remark this answers is not in the document]
Comment (tsimbalar) [reply src/Serilog.Settings.Configuration/Settings/Configuration/Assemblies/DllScanningAssemblyFinder.cs]: I've created https://github.com/serilog/serilog-settings-configuration/issues/157 to track the issue that can be solved in a separate PR.

File: src/Serilog.Settings.Configuration/Settings/Configuration/Assemblies/AssemblyFinder.cs [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+        protected static bool IsCaseInsensitiveMatch(string text, string textToFind)
+        {
+            return text != null && text.ToLowerInvariant().Contains(textToFind.ToLowerInvariant());
+        }
+
+        public static AssemblyFinder Auto()
+        {
+            if (Assembly.GetEntryAssembly() != null)
+            {
```
[the remark this answers is not in the document]
Comment (tsimbalar) [reply src/Serilog.Settings.Configuration/Settings/Configuration/Assemblies/AssemblyFinder.cs]: Done !

File: src/Serilog.Settings.Configuration/Settings/Configuration/Assemblies/AssemblyFinder.cs [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+                return new DependencyContextAssemblyFinder(DependencyContext.Default);
```
[the remark this answers is not in the document]
Comment (tsimbalar) [reply src/Serilog.Settings.Configuration/Settings/Configuration/Assemblies/AssemblyFinder.cs]: Done !

File: src/Serilog.Settings.Configuration/Settings/Configuration/ResolutionContext.cs [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    /// Keeps track of available elements that are useful when resolving values in the settings system.
+    /// </summary>
+    sealed class ResolutionContext
+    {
+        readonly IDictionary<string, LoggingLevelSwitch> _declaredLevelSwitches;
+        readonly IConfiguration _appConfiguration;
+
+        public ResolutionContext(IConfiguration appConfiguration = null)
+        {
+            _declaredLevelSwitches = new Dictionary<string, LoggingLevelSwitch>();
```
[the remark this answers is not in the document]
Comment (tsimbalar) [reply src/Serilog.Settings.Configuration/Settings/Configuration/ResolutionContext.cs]: After reviewing the code, I see that we are not using any *case insensitive comparer* anywhere where we initialize `Dictionary`s ... I'd rather not do the change (at least not in this PR), in order not to introduce any subtle bugs. 

Shall we open a separate issue to keep track of it ?

File: src/Serilog.Settings.Configuration/Settings/Configuration/ConfigurationReader.cs [hunk 5]
```diff
@@ -265,11 +253,11 @@ string GetSectionName(IConfigurationSection s)
             }
         }
 
-        IReadOnlyCollection<Assembly> LoadConfigurationAssemblies()
+        static IReadOnlyCollection<Assembly> LoadConfigurationAssemblies(IConfigurationSection section, AssemblyFinder assemblyFinder)
         {
             var assemblies = new Dictionary<string, Assembly>();
```
[the remark this answers is not in the document]
Comment (tsimbalar) [reply src/Serilog.Settings.Configuration/Settings/Configuration/ConfigurationReader.cs]: same comment as above : 

> After reviewing the code, I see that we are not using any *case insensitive comparer* anywhere where we initialize `Dictionary`s ... I'd rather not do the change (at least not in this PR), in order not to introduce any subtle bugs. 
> 
> Shall we open a separate issue to keep track of it ?

**Comment by skomis-mm** [comment]
@tsimbalar Thanks! Let push this to the dev to try this in the wild 👍

**Pull request merged**